Mathlib/Order/SuccPred/Basic.lean

```diff
@@ -1401,3 +1401,35 @@ lemma SuccOrder.forall_ne_bot_iff
   rw [← Nat.succ_pred_eq_of_pos hj]
   simp only [Function.iterate_succ', Function.comp_apply]
   apply h
+
+section IsLeast
+
+lemma BddAbove.exists_isGreatest_of_nonempty {X : Type*} [LinearOrder X] [SuccOrder X]
+    [IsSuccArchimedean X] {S : Set X} (hS : BddAbove S) (hS' : S.Nonempty) :
+    ∃ x, IsGreatest S x := by
```

Comment:
@YaelDillies Are these the right typeclass assumptions?

Reply:
I guess the right assumption is something like "locally cowell-founded", meaning "`Iic x` is cowell-founded for all `x`". We don't have that in mathlib, so the current generality seems okay.

Note however that `LinearOrder` shouldn't be necessary since the recent refactor on `SuccOrder`
```suggestion
lemma BddAbove.exists_isGreatest_of_nonempty {X : Type*} [Preorder X] [SuccOrder X]
    [IsSuccArchimedean X] {S : Set X} (hS : BddAbove S) (hS' : S.Nonempty) :
    ∃ x, IsGreatest S x := by
```

Reply:
I can't generalize to `Preorder` because I use `Order.le_succ_iff_eq_or_le` in the inductive step proof of `Succ.rec`.

Reply:
The end result is still true, though (although I've noticed you also need to assume that the set is directed). Can you try using another lemma?

Reply:
@Command-Master, didn't you prove that the assumptions above imply `SemilatticeSup X`?

Reply:
Should this now use results from #16559?

Reply:
There was no new result in #16559. Do you mean #16272?

Reply:
Yes, I now see the results got split

Reply:
I've left a TODO to come back to this after that PR.

Reply:
Sorry, I won't merge this for now because I suspect you don't even need the lemmas in the current PR. Please first coordinate with @Command-Master to get #16272 (or rather the preliminary part) merged.

Reply:
Can you take a look at https://github.com/leanprover-community/mathlib4/pull/16619/files#diff-e9c6f03b09ef6fa37d4bb4f98167c69a8165fcbbd8f189451b15fafbf798f65cR113 please? It's the point of use.

Reply:
I tried to bring in the relevant lemmas from that PR, but I still don't have a proof. I also can't assume that my set is `OrdConnected`. I need help with this.